master/internal/core_checkpoint.go

```diff
@@ -0,0 +1,397 @@
+
+	return n, err
+}
+
+// BatchDownloadIterator implements s3's BatchDownloadIterator API.
+type batchDownloadIterator struct {
+	// The objects we are writing
+	objects []*s3.Object
+	// The output we are writing to
+	aw archiveWriter
+	// Internal states
+	err    error
+	pos    int
+	bucket string
+}
+
+// Next() returns true if the next item is available.
+func (i *batchDownloadIterator) Next() bool {
+	i.pos++
+	if i.pos == len(i.objects) {
+		return false
+	}
+	err := i.aw.WriteFileHeader(*i.objects[i.pos].Key, *i.objects[i.pos].Size)
+	if err != nil {
+		i.err = err
+		return false
+	}
+	return true
+}
+
+// Err() eturns the error if any.
+func (i *batchDownloadIterator) Err() error {
+	return i.err
+}
+
+// DownloadObject() eturns a DownloadObject.
+func (i *batchDownloadIterator) DownloadObject() s3manager.BatchDownloadObject {
+	return s3manager.BatchDownloadObject{
+		Object: ptrs.Ptr(s3.GetObjectInput{
+			Bucket: ptrs.Ptr(i.bucket),
+			Key:    i.objects[i.pos].Key,
+		}),
+		Writer: newS3SeqWriterAt(i.aw),
+	}
+}
+
+func newBatchDownloadIterator(aw archiveWriter,
+	bucket string, objs []*s3.Object) *batchDownloadIterator {
+	return &batchDownloadIterator{
+		aw:      aw,
+		bucket:  bucket,
+		objects: objs,
+		pos:     -1,
+	}
+}
+
+func getAwsRegion() string {
+	defaultRegion := "us-west-2"
+	client := http.Client{
+		Timeout: 100 * time.Millisecond,
+	}
+
+	// Per AWS: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+	resp, err := client.Get("http://169.254.169.254/latest/dynamic/instance-identity/document")
+	if err != nil || resp.StatusCode != http.StatusOK {
+		return defaultRegion
+	}
+
+	bytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return defaultRegion
+	}
+	_ = resp.Body.Close()
+
+	var jsonObj map[string]interface{}
+	err = json.Unmarshal(bytes, &jsonObj)
+	if err != nil {
+		return defaultRegion
+	}
+
+	if _, ok := jsonObj["region"]; !ok {
+		return defaultRegion
+	}
+
+	region, ok := jsonObj["region"].(string)
+	if !ok {
+		return defaultRegion
+	}
+
+	return region
+}
+
+func s3DownloadCheckpoint(
+	c echo.Context, aw archiveWriter, bucket string, prefix string,
+) error {
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(getAwsRegion()),
+	})
+	if err != nil {
+		return err
+	}
+	// We do not pass in credentials explicitly. Instead, we reply on
+	// the existing AWS credentials.
+	s3client := s3.New(sess)
+
+	var errs []error
+	downloader := s3manager.NewDownloader(sess, func(d *s3manager.Downloader) {
+		d.Concurrency = 1 // Setting concurrency to 1 to use s3SeqWriterAt
+	})
+	funcReadPage := func(output *s3.ListObjectsV2Output, lastPage bool) bool {
+		iter := newBatchDownloadIterator(aw, bucket, output.Contents)
+		// Download every bucket in this page
+		err = downloader.DownloadWithIterator(c.Request().Context(), iter)
+		if iter.Err() != nil {
+			errs = append(errs, iter.Err())
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
```

Comment:
Question about this code: if `iter.Err()` is set, is it possible for DownloadWithIterator to return any value other than that error?

Reply:
Yeah, `iter.Err()` isn't called in [DownloadWithIterator](https://github.com/aws/aws-sdk-go/blob/2ab2a7fda76a11d2df0a0d15e554de2139672d31/service/s3/s3manager/download.go#L259), hence I need to record the error here.

Reply:
oh fantastic, I was curious where that implementation was.

That's also worth explaining in a comment, imo.

Reply:
You can just click the link in my last comment.